source/common/config/version_converter.cc

```diff
@@ -0,0 +1,123 @@
+      for (int n = 0; n < field_size; ++n) {                                                       \
+        const auto& v =                                                                            \
+            prev_reflection->GetRepeated##method_fragment(prev_message, prev_field_descriptor, n); \
+        target_reflection->Add##method_fragment(target_message, target_field_descriptor, v);       \
+      }                                                                                            \
+    } else {                                                                                       \
+      const auto v = prev_reflection->Get##method_fragment(prev_message, prev_field_descriptor);   \
+      target_reflection->Set##method_fragment(target_message, target_field_descriptor, v);         \
+    }                                                                                              \
+    break;                                                                                         \
+  }
+
+namespace Envoy {
+namespace Config {
+
+constexpr uint32_t DeprecatedMessageFieldNumber = 100000;
+
+void VersionConverter::upgrade(const Protobuf::Message& prev_message,
+                               Protobuf::Message& next_message) {
+  const Protobuf::Descriptor* next_descriptor = next_message.GetDescriptor();
+  const Protobuf::Reflection* prev_reflection = prev_message.GetReflection();
+  std::vector<const Protobuf::FieldDescriptor*> prev_field_descriptors;
+  prev_reflection->ListFields(prev_message, &prev_field_descriptors);
+  Protobuf::DynamicMessageFactory dmf;
+  std::unique_ptr<Protobuf::Message> deprecated_message;
+
+  // Iterate over all the set fields in the previous version message.
+  for (const auto* prev_field_descriptor : prev_field_descriptors) {
+    const Protobuf::Reflection* target_reflection = next_message.GetReflection();
+    Protobuf::Message* target_message = &next_message;
+
+    // Does the field exist in the new version message?
+    const std::string& prev_name = prev_field_descriptor->name();
+    const auto* target_field_descriptor = next_descriptor->FindFieldByName(prev_name);
+    // If we can't find this field in the next version, it must be deprecated.
+    // So, use deprecated_message and its reflection instead.
+    if (target_field_descriptor == nullptr) {
+      ASSERT(prev_field_descriptor->options().deprecated());
+      if (!deprecated_message) {
+        deprecated_message.reset(dmf.GetPrototype(prev_message.GetDescriptor())->New());
+      }
+      target_field_descriptor = prev_field_descriptor;
+      target_reflection = deprecated_message->GetReflection();
+      target_message = deprecated_message.get();
+    }
+    ASSERT(target_field_descriptor != nullptr);
+
+    // These properties are guaranteed by protoxform.
+    ASSERT(prev_field_descriptor->type() == target_field_descriptor->type());
```

Comment:
Is there anyway we can check whether the descriptor of next_message is translated by protoxform from that of prev_message? This seems generally true for configs in this repo but not outside.

Reply:
I think this will be done by code outside of the version converter.. e.g. the HCM filter factory instantiation will need to be explicitly aware of the `envoy.foo.v2`, `envoy.foo.v3alpha`, `envoy.foo.v4` structure to be able to stage a series of upgrades anyway, so there will be some additional code to recognize types that are in the core Envoy API/UDPA and interpret the version numbers.